pkg/vmprovider/providers/vsphere2/session/session_vm_update.go

```diff
@@ -300,59 +300,111 @@ func UpdateConfigSpecExtraConfig(
 	vmClassSpec *vmopv1.VirtualMachineClassSpec,
 	vm *vmopv1.VirtualMachine,
 	globalExtraConfig map[string]string,
-	imageV1Alpha1Compatible bool) {
+	imgV1A1Compat bool) {
 
-	extraConfig := maps.Clone(globalExtraConfig)
+	// Either initialize extraConfig to an empty map or from a copy of
+	// globalExtraConfig.
+	var extraConfig map[string]string
+	if globalExtraConfig == nil {
+		extraConfig = map[string]string{}
+	} else {
+		extraConfig = maps.Clone(globalExtraConfig)
+	}
 
-	virtualDevices := vmClassSpec.Hardware.Devices
-	if len(virtualDevices.VGPUDevices) > 0 || len(virtualDevices.DynamicDirectPathIODevices) > 0 ||
-		(classConfigSpec != nil && util.HasVirtualPCIPassthroughDeviceChange(classConfigSpec.DeviceChange)) {
-		setMMIOExtraConfig(vm, extraConfig)
+	// Ensure a VM with vGPUs or dynamic direct path I/O devices has the the
+	// correct flags in ExtraConfig for memory mapped I/O. Please see
+	// https://kb.vmware.com/s/article/2142307 for more information about these
+	// flags.
+	if hasvGPUOrDDPIODevices(classConfigSpec, vmClassSpec) {
```

Comment:
If we're doing this while the VM is on shouldn't it really be looking at the VM's ConfigInfo instead of the class?

Reply:
Technically if the VM is deployed at all. I guess the current code path is an artifact of a single update post deploy.